### PR TITLE
*: add start-relay/stop-relay command

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
add `start-relay` and `stop-relay`, where we deprecate `EnableRelay` in source config files.

TODO (will file another final PR): 
1. add a command to query relay 
2. send `purge-relay` to all relay workers

### What is changed and how it works?
let `start-relay`, `stop-relay` take effect, DM worker should call `EnableRelay` and `EnableHandleSubtask` accordingly.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has persistent data change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### CODE REVIEW

**Comment by lance6716** [comment]
waiting for https://github.com/pingcap/dm/pull/1508 https://github.com/pingcap/dm/pull/1511

**Comment by lance6716** [comment]
log flood with 

```
[2021-03-18T11:50:31.398Z] [2021/03/18 19:50:01.729 +08:00] [WARN] [reader.go:545] ["fail to parse relay log file, meet some ignorable error"] [task=test] [unit="binlog replication"] [component="binlog reader"] [file=relay-dir/81dc0587-87df-11eb-8a5c-725d17e230d2.000001/test-ee2de0cc-739d-4c21-9754-513c1e087159-6gv88-0g3vt-bin.000005] [offset=1028] [error="get event err EOF, need 16777216 but got 304"] [errorVerbose="get event err EOF, need 16777216 but got 304\ngithub.com/siddontang/go-mysql/replication.(*BinlogParser).parseSingleEvent\n\t/go/pkg/mod/github.com/lance6716/go-mysql@v1.1.1-0.20210303100354-b0e44c2c5623/replication/parser.go:130\ngithub.com/siddontang/go-mysql/replication.(*BinlogParser).ParseReader\n\t/go/pkg/mod/github.com/lance6716/go-mysql@v1.1.1-0.20210303100354-b0e44c2c5623/replication/parser.go:166\ngithub.com/siddontang/go-mysql/replication.(*BinlogParser).ParseFile\n\t/go/pkg/mod/github.com/lance6716/go-mysql@v1.1.1-0.20210303100354-b0e44c2c5623/replication/parser.go:96\ngithub.com/pingcap/dm/pkg/streamer.(*BinlogReader).parseFile\n\t/home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/streamer/reader.go:542\ngithub.com/pingcap/dm/pkg/streamer.(*BinlogReader).parseFileAsPossible\n\t/home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/streamer/reader.go:411\ngithub.com/pingcap/dm/pkg/streamer.(*BinlogReader).parseDirAsPossible\n\t/home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/streamer/reader.go:379\ngithub.com/pingcap/dm/pkg/streamer.(*BinlogReader).parseRelay\n\t/home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/streamer/reader.go:308\ngithub.com/pingcap/dm/pkg/streamer.(*BinlogReader).StartSyncByGTID.func1\n\t/home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/streamer/reader.go:281\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357"]
```

**Comment by lance6716** [comment]
TODO: move `setSourceStatus` closer to `EnableXXX`

File: dm/master/server.go [hunk 1]
```diff
@@ -1126,6 +1126,13 @@ func (s *Server) OperateSource(ctx context.Context, req *pb.OperateSourceRequest
 	}
 
 	resp.Result = true
+	// tell user he should use `start-relay` to manually specify relay workers
+	for _, cfg := range cfgs {
+		if cfg.EnableRelay {
+			resp.Msg = "Please use `start-relay` to specify which workers should pull relay log of relay-enabled sources."
```
Comment (GMHDBJD) [inline dm/master/server.go:1132]: It's confused for the user if they don't need to start-relay in multiple workers.

File: dm/config/source_config.go [hunk 2]
```diff
@@ -177,17 +177,15 @@ func (c *SourceConfig) Verify() error {
...
-			if err != nil {
-				return terror.WithClass(terror.Annotatef(err, "relay-binlog-gtid %s", c.RelayBinlogGTID), terror.ClassDMWorker)
-			}
+	}
+	if len(c.RelayBinlogGTID) > 0 {
+		_, err = gtid.ParserGTID(c.Flavor, c.RelayBinlogGTID)
+		if err != nil {
+			return terror.WithClass(terror.Annotatef(err, "relay-binlog-gtid %s", c.RelayBinlogGTID), terror.ClassDMWorker)
 		}
 	}
```
[the remark this answers is not in the document]
Comment (lance6716) [reply dm/config/source_config.go:190]: they may start relay for this source later, so we must check at this time

File: dm/master/server.go (hunk 1 above)
Comment (lance6716) [reply dm/master/server.go:1132]: for current behaviour, they must using `start-relay` to pull relay log. This's the only change to tell uesr we changed the behaviour

File: dm/ctl/master/start_relay.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		}
+		err = errors.New("please check output to see error")
+		return
+	}
+
+	if len(sources) != 1 {
+		common.PrintLines("must specify one source (`-s` / `--source`)")
+		err = errors.New("please check output to see error")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (lance6716) [reply dm/ctl/master/start_relay.go:59]: to keep same behaviour with other commands

https://github.com/pingcap/dm/tree/master/tests/dmctl_basic/check_list

File: dm/config/source_config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lance6716) [reply dm/config/source_config.go:190]: Oh I see

File: dm/master/scheduler/scheduler.go [hunk 4]
```diff
@@ -1326,7 +1546,21 @@ func (s *Scheduler) tryBoundForSource(source string) (bool, error) {
 			}
 		}
 	}
-
+	// then a relay worker for this source...
+	if worker == nil {
+		for workerName := range s.relayWorkers[source] {
+			w, ok := s.workers[workerName]
+			if !ok {
+				// a not found worker, should not happened
```
Comment (GMHDBJD) [inline dm/master/scheduler/scheduler.go:1554]: what does this mean

**Review by lichunzhu — COMMENTED** [review]
/lgtm

**Pull request merged**